### PR TITLE
Feature/DOCI

### PR DESCRIPTION
**Pull request by tmhuysen**

### CODE REVIEW

File: include/HamiltonianBuilder/DOCI.hpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+#include <memory>
+
+
+
+namespace GQCG {
+
+
+/**
+ *  Doubly occupied configuration interaction builds a hamiltonian matrix
+ *  based on a wavefunction only containing doubly occupied configurations
```
Comment (lelemmen) [inline include/HamiltonianBuilder/DOCI.hpp]: add punctiation .

File: include/HamiltonianBuilder/DOCI.hpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ *  This requires the alpha and beta Fock space to be identical
+ *  combinations of identical alpha, beta ONVs are used.
```
Comment (lelemmen) [inline include/HamiltonianBuilder/DOCI.hpp]: I feel like this sentence is unnecessary

File: include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+
+
+namespace GQCG {
+
+
+/**
+ *  RestrictedHamiltonianBuilder is an abstract base class for quantum chemical methods performed with restricted hamiltonian parameters
+ *  for which the Hamiltonian is preferably expressed as a Hermitian matrix
+ *  so that the corresponding eigen-values and -vectors can be retrieved through diagonalisation of this matrix.
```
Comment (lelemmen) [inline include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp]: 'eigenvalues' and -vectors

File: include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ */
+class RestrictedHamiltonianBuilder {
+protected:
+    HamiltonianParameters hamiltonian_parameters;  // the hamiltonian parameters passed for the calculations
```
Comment (lelemmen) [inline include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp]: Hamiltonian

File: include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#include <memory>
+#include <utility>
+
+
+
+namespace GQCG {
+
+
+/**
+ *  RestrictedHamiltonianBuilder is an abstract base class for quantum chemical methods performed with restricted hamiltonian parameters
```
Comment (lelemmen) [inline include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp]: Hamiltonian

File: include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ *  for which the Hamiltonian is preferably expressed as a Hermitian matrix
+ *  so that the corresponding eigen-values and -vectors can be retrieved through diagonalisation of this matrix.
+ */
+class RestrictedHamiltonianBuilder {
+protected:
+    HamiltonianParameters hamiltonian_parameters;  // the hamiltonian parameters passed for the calculations
+
+    Eigen::VectorXd diagonal;  // the diagonal of the hamiltonian matrix
```
Comment (lelemmen) [inline include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp]: Hamiltonian

File: include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+
+    // PROTECTED CONSTRUCTORS
+    /**
+     *  Protected constructor given a @param HamiltonianParameters
```
Comment (lelemmen) [inline include/HamiltonianBuilder/RestrictedHamiltonianBuilder.hpp]: `hamiltonian_parameters`

File: include/Operator/OneElectronOperator.hpp [hunk 8]
```diff
@@ -29,7 +29,8 @@ class OneElectronOperator : public BaseOperator {
 
     // GETTERS
     Eigen::MatrixXd get_matrix_representation() const { return this->matrix; }
-    
+    double get(size_t p, size_t q) { return this->matrix(p,q); }
```
Comment (lelemmen) [inline include/Operator/OneElectronOperator.hpp]: We might (anticipating on abstracting the effective representation of the operators) also implement `operator()`. Can you make an issue out of this?

File: include/Operator/TwoElectronOperator.hpp [hunk 9]
```diff
@@ -22,6 +22,7 @@ class TwoElectronOperator : public BaseOperator {
 public:
     // GETTERS
     Eigen::Tensor<double, 4> get_matrix_representation() const { return this->tensor; }
+    double get(size_t p, size_t q, size_t r, size_t s) { return this->tensor(p, q, r, s); }
```
Comment (lelemmen) [inline include/Operator/TwoElectronOperator.hpp]: see comment in OneElectronOperator.hpp

File: src/HamiltonianBuilder/DOCI.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+/*
+ *  CONSTRUCTORS
+ */
+
+/**
+ *  Constructor given a @param hamiltonian_parameters and @param fock_space
+ */
+DOCI::DOCI(HamiltonianParameters hamiltonian_parameters, FockSpace fock_space):
+    RestrictedHamiltonianBuilder(hamiltonian_parameters),
```
Comment (lelemmen) [inline src/HamiltonianBuilder/DOCI.cpp]: add whitespace

File: src/HamiltonianBuilder/DOCI.cpp [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    fock_space(fock_space),
```
Comment (lelemmen) [inline src/HamiltonianBuilder/DOCI.cpp]: add whitespace

File: src/HamiltonianBuilder/DOCI.cpp [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    dim(fock_space.get_dimension())
```
Comment (lelemmen) [inline src/HamiltonianBuilder/DOCI.cpp]: add whitespace

File: src/HamiltonianBuilder/DOCI.cpp [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+{
+    auto K = this->hamiltonian_parameters.get_h().get_dim();
+    if (K != this->fock_space.K) {
+        throw std::invalid_argument("Basis functions of the Fock space and AObasis are incompatible.");
```
Comment (lelemmen) [inline src/HamiltonianBuilder/DOCI.cpp]: AOBasis

File: src/HamiltonianBuilder/DOCI.cpp [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+                    onv.annihilate(p);
+                    onv.create(q);
+
+                    size_t J = this->fock_space.getAddress(onv);  // J is the address of a string that couples to I
+
+                    // The loops are p->K and q<p. So, we should normally multiply by a factor 2 (since the summand is symmetric)
+                    // However, we are setting both of the symmetric indices of Hamiltonian, so no factor 2 is required
+                    matvec(I) += this->hamiltonian_parameters.get_g().get(p, q, p, q)* x(J);
+                    matvec(J) += this->hamiltonian_parameters.get_g().get(p, q, p, q)* x(I);
```
Comment (lelemmen) [inline src/HamiltonianBuilder/DOCI.cpp]: add whitespace

File: src/HamiltonianBuilder/DOCI.cpp [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+                if (!onv.isOccupied(q)) {  // if q not in I
+
+                    onv.annihilate(p);
+                    onv.create(q);
+
+                    size_t J = this->fock_space.getAddress(onv);  // J is the address of a string that couples to I
+
+                    // The loops are p->K and q<p. So, we should normally multiply by a factor 2 (since the summand is symmetric)
+                    // However, we are setting both of the symmetric indices of Hamiltonian, so no factor 2 is required
+                    matvec(I) += this->hamiltonian_parameters.get_g().get(p, q, p, q)* x(J);
```
Comment (lelemmen) [inline src/HamiltonianBuilder/DOCI.cpp]: add whitespace

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    g_tensor.setRandom();
+    GQCG::TwoElectronOperator g (g_tensor);
+    Eigen::MatrixXd C = Eigen::MatrixXd::Random(K, K);
+    GQCG::HamiltonianParameters random_hamiltonian_parameters (ao_basis_ptr, S, H_core, g, C);
+
+    // Create a compatible Fock space
+    GQCG::FockSpace fock_space (K, 3);
+
+    // Check if a correct constructor works and public methods don't fail
+    BOOST_CHECK_NO_THROW(GQCG::DOCI random_doci(random_hamiltonian_parameters, fock_space));
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: no need for giving a name to an instance inside `BOOST_CHECK_...`

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+
+    // Check if faulty constructor parameters throw an error
+    // Create an incompatible Fock space
+    GQCG::FockSpace fock_space_2 (K+1, 3);
+    BOOST_CHECK_THROW(GQCG::DOCI random_doci_2(random_hamiltonian_parameters,fock_space_2), std::invalid_argument);
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: no need for giving a name to an instance inside `BOOST_CHECK_...`

Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: add whitespace in constructor arguments

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+
+#include <boost/test/unit_test.hpp>
+#include <boost/test/included/unit_test.hpp>  // include this to get main(), otherwise the compiler will complain
+
+
+
+BOOST_AUTO_TEST_CASE ( DOCI_constructor  ) {
+    // Create an AOBasis
+    GQCG::Molecule water ("../tests/data/h2o.xyz");
+    auto ao_basis_ptr = std::make_shared<GQCG::AOBasis>(water, "STO-3G");
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: remove '_sptr' in names

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+    // Create an incompatible Fock space
+    GQCG::FockSpace fock_space_2 (K+1, 3);
+    BOOST_CHECK_THROW(GQCG::DOCI random_doci_2(random_hamiltonian_parameters,fock_space_2), std::invalid_argument);
+}
+
+
+BOOST_AUTO_TEST_CASE ( DOCI_public_methods ) {
+    // Create an AOBasis
+    GQCG::Molecule water ("../tests/data/h2o.xyz");
+    auto ao_basis_ptr = std::make_shared<GQCG::AOBasis>(water, "STO-3G");
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: remove '_sptr' in names

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+
+    // Create random HamiltonianParameters from One- and TwoElectronOperators (and a transformation matrix) with compatible dimensions
+    size_t K = ao_basis_ptr->get_number_of_basis_functions();
+    GQCG::OneElectronOperator S (Eigen::MatrixXd::Random(K, K));
+    GQCG::OneElectronOperator H_core (Eigen::MatrixXd::Random(K, K));
+    Eigen::Tensor<double, 4> g_tensor (K, K, K, K);
+    g_tensor.setRandom();
+    GQCG::TwoElectronOperator g (g_tensor);
+    Eigen::MatrixXd C = Eigen::MatrixXd::Random(K, K);
+    GQCG::HamiltonianParameters random_hamiltonian_parameters (ao_basis_ptr, S, H_core, g, C);
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: remove '_sptr' in names

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+
+
+BOOST_AUTO_TEST_CASE ( DOCI_public_methods ) {
+    // Create an AOBasis
+    GQCG::Molecule water ("../tests/data/h2o.xyz");
+    auto ao_basis_ptr = std::make_shared<GQCG::AOBasis>(water, "STO-3G");
+
+
+    // Create random HamiltonianParameters from One- and TwoElectronOperators (and a transformation matrix) with compatible dimensions
+    size_t K = ao_basis_ptr->get_number_of_basis_functions();
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: remove '_sptr' in names

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+    GQCG::OneElectronOperator S (Eigen::MatrixXd::Random(K, K));
+    GQCG::OneElectronOperator H_core (Eigen::MatrixXd::Random(K, K));
+    Eigen::Tensor<double, 4> g_tensor (K, K, K, K);
+    g_tensor.setRandom();
+    GQCG::TwoElectronOperator g (g_tensor);
+    Eigen::MatrixXd C = Eigen::MatrixXd::Random(K, K);
+    GQCG::HamiltonianParameters random_hamiltonian_parameters (ao_basis_ptr, S, H_core, g, C);
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: remove '_sptr' in names

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+
+    // Create a compatible Fock space
+    GQCG::FockSpace fock_space (K, 3);
+
+    // Create DOCI module
+    GQCG::DOCI random_doci (random_hamiltonian_parameters,fock_space);
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: add whitespace in constructor arguments

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+
+#include "HamiltonianBuilder/DOCI.hpp"
+#include "HamiltonianParameters/HamiltonianParameters_constructors.hpp"
+
+#include <boost/test/unit_test.hpp>
+#include <boost/test/included/unit_test.hpp>  // include this to get main(), otherwise the compiler will complain
+
+
+
+BOOST_AUTO_TEST_CASE ( DOCI_constructor  ) {
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: remove excessive whitespace in test name

File: tests/HamiltonianBuilder/DOCI_test.cpp [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+    Eigen::Tensor<double, 4> g_tensor (K, K, K, K);
+    g_tensor.setRandom();
+    GQCG::TwoElectronOperator g (g_tensor);
+    Eigen::MatrixXd C = Eigen::MatrixXd::Random(K, K);
+    GQCG::HamiltonianParameters random_hamiltonian_parameters (ao_basis_ptr, S, H_core, g, C);
+
+    // Create a compatible Fock space
+    GQCG::FockSpace fock_space (K, 3);
+
+    // Check if a correct constructor works and public methods don't fail
```
Comment (lelemmen) [inline tests/HamiltonianBuilder/DOCI_test.cpp]: public methods aren't tested in this test, remove the comment

File: include/HamiltonianBuilder/DOCI.hpp (hunk 2 above)
Comment (tmhuysen) [reply include/HamiltonianBuilder/DOCI.hpp]: Maybe it is poorly worded, but with two identical Fock spaces you can do FCI when you let the alpha and beta ONVs vary independently.

File: include/Operator/OneElectronOperator.hpp (hunk 8 above)
Comment (tmhuysen) [reply include/Operator/OneElectronOperator.hpp]: I don't understand your request.

File: src/HamiltonianBuilder/DOCI.cpp (hunk 13 above)
Comment (tmhuysen) [reply src/HamiltonianBuilder/DOCI.cpp]: FCIDUMP files do not have an AOBasis for example.

File: src/HamiltonianBuilder/DOCI.cpp (hunk 15 above)
Comment (tmhuysen) [reply src/HamiltonianBuilder/DOCI.cpp]: where?

File: include/HamiltonianBuilder/DOCI.hpp (hunk 2 above)
Comment (lelemmen) [reply include/HamiltonianBuilder/DOCI.hpp]: indeed, word this differently

File: src/HamiltonianBuilder/DOCI.cpp (hunk 15 above)
Comment (lelemmen) [reply src/HamiltonianBuilder/DOCI.cpp]: `matvec(I) += this->hamiltonian_parameters.get_g().get(p, q, p, q) * x(J)`

File: tests/HamiltonianBuilder/DOCI_test.cpp (hunk 25 above)
Comment (lelemmen) [reply tests/HamiltonianBuilder/DOCI_test.cpp]: `// Check if a correct constructor works`

**Pull request merged**